internal/cose.go

```diff
@@ -0,0 +1,165 @@
+package internal
+
+import (
+	"crypto/ecdsa"
+	"crypto/sha512"
+	"fmt"
+	"math/big"
+
+	"github.com/fxamacker/cbor/v2"
+)
+
+type SigningAlgorithm int
+
+const (
+	BadOrMissingAlgorithm SigningAlgorithm = iota
+	ES384
+)
+
+// https://datatracker.ietf.org/doc/html/rfc8152#section-4.2
+type CoseSign1 struct {
+	_ struct{} `cbor:",toarray"`
+
+	Protected   []byte
+	Unprotected cbor.RawMessage
+	Payload     []byte
+	Signature   []byte
+}
+
+type CoseHeader struct {
+	Alg interface{} `cbor:"1,keyasint,omitempty" json:"alg,omitempty"`
+}
+
+func MakeCoseSign1FromBytes(bytes []byte) (CoseSign1, error) {
+	coseSign1 := CoseSign1{}
+	err := cbor.Unmarshal(bytes, &coseSign1)
+	if nil != err {
+		return CoseSign1{}, fmt.Errorf("unmarshaling CoseSign1 from bytes: %w", err)
+	}
+
+	if len(coseSign1.Protected) == 0 {
+		return CoseSign1{}, fmt.Errorf("missing cose protected headers")
+	}
+
+	if len(coseSign1.Payload) == 0 {
+		return CoseSign1{}, fmt.Errorf("missing cose payload")
+	}
+
+	if len(coseSign1.Signature) == 0 {
+		return CoseSign1{}, fmt.Errorf("missing cose signature")
+	}
+	return coseSign1, nil
+}
+
+func (c CoseSign1) CheckSignature(
+	publicKey *ecdsa.PublicKey,
+) ([]byte, error) {
+	signingAlg, err := c.GetSigningAlgorithm()
+	if err != nil {
+		return nil, fmt.Errorf("extracting signing algorithm: %w", err)
+	}
+
+	// https://datatracker.ietf.org/doc/html/rfc8152#section-4.4
+	sigStruct := struct {
+		_ struct{} `cbor:",toarray"`
+
+		Context     string
+		Protected   []byte
+		ExternalAAD []byte
+		Payload     []byte
+	}{
+		Context:     "Signature1",
+		Protected:   c.Protected,
+		ExternalAAD: []byte{},
+		Payload:     c.Payload,
+	}
+
+	sigStructBytes, err := cbor.Marshal(&sigStruct)
+	if err != nil {
+		return nil, fmt.Errorf("marshaling cose signature struct: %w", err)
+	}
+
+	var sigStructHash []byte
+	switch signingAlg {
+	case ES384:
+		if publicKey.Curve.Params().Name != "P-384" {
+			return nil, fmt.Errorf(
+				"cose signing alg '%v' does not match public key signing alg '%s'",
+				signingAlg,
+				publicKey.Curve.Params().Name,
+			)
+		}
+
+		h := sha512.Sum384(sigStructBytes)
+		sigStructHash = h[:]
+	default:
+		return nil, fmt.Errorf("unsupported signing alg '%v'", signingAlg)
```

Comment:
Does this case need a test?

Reply:
It's not actually reachable under the current implementation as the `GetSigningAlgorithm()` function only supports `ES384` so this type of "unsupported signing alg" gets caught above. I still need a default case for the switch statement, however, and that case would be an unsupported signing algo.